exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java

```diff
@@ -1,169 +1,252 @@
+        } else {
+
+            // TODO(AR) could the class member `map` remain `linear` ?
+
+            // create a transient map
+            final IMap<AtomicValue, Sequence> newMap = map.linear();
+
+            for (final IEntry<AtomicValue, Sequence> entry : other) {
+                newMap.put(entry.key(), entry.value());
+            }
+
+            // return to immutable map
+            map = newMap.forked();
+        }
+    }
+
+    @Override
+    public AbstractMapType merge(final Iterable<AbstractMapType> others) {
+
+        // create a transient map
+        IMap<AtomicValue, Sequence> newMap = map.linear();
+
+        int prevType = type;
+        for (final AbstractMapType other: others) {
+            if (other instanceof MapType) {
+                // MapType - optimise merge
+                final MapType otherMap = (MapType) other;
+                newMap = map.union(otherMap.map);
+
+                if (prevType != otherMap.type) {
+                    prevType = Type.ITEM;
                 }
-                map = tmap.persistentMap();
             } else {
-                for (final Map.Entry<AtomicValue, Sequence> entry : other)
-                    map = map.assoc(entry.getKey(), entry.getValue());
+                // non MapType
+                for (final IEntry<AtomicValue, Sequence> entry : other) {
+                    final AtomicValue key = entry.key();
+                    newMap = newMap.put(key, entry.value());
+                    if (prevType != key.getType()) {
+                        prevType = Type.ITEM;
+                    }
+                }
             }
         }
+
+        // return an immutable map
+        return new MapType(context, newMap.forked(), prevType);
     }
 
-    public void add(AtomicValue key, Sequence value) {
+    public void add(final AtomicValue key, final Sequence value) {
         setKeyType(key.getType());
-        this.map = this.map.assoc(key, value);
+        map = map.put(key, value);
     }
 
+    @Override
     public Sequence get(AtomicValue key) {
         key = convert(key);
-        if (key == null)
-            {return Sequence.EMPTY_SEQUENCE;}
-        final Map.Entry<AtomicValue, Sequence> e = this.map.entryAt(key);
-        return e == null ? Sequence.EMPTY_SEQUENCE : e.getValue();
+        if (key == null) {
+            return Sequence.EMPTY_SEQUENCE;
+        }
+
+        final Sequence result = map.get(key, null);
+        return result == null ? Sequence.EMPTY_SEQUENCE : result;
     }
 
     @Override
-    public AbstractMapType put(AtomicValue key, final Sequence value) throws XPathException {
-        return new MapType(this.context, this.map.assoc(key, value), type);
+    public AbstractMapType put(final AtomicValue key, final Sequence value) {
+        final IMap<AtomicValue, Sequence> newMap = map.put(key, value);
+        return new MapType(this.context, newMap, type == key.getType() ? type : Type.ITEM);
```

Comment:
In what case does the key not have a type? And shouldn't it be `Type.ATOMIC` everywhere? 
Couldn't `Type.ITEM` be a node or element, which would not qualify for a map key?

Reply:
Right. I just followed the type stuff that was previously there. That is too broad, I have now narrowed the types as you suggested.